[sdk/go] Panic when calling a method on a resource that implements `pulumi.Input`

We're seeing a panic when calling a method on a resource that implements `pulumi.Input`, which will be the case for all code-generated resources.

```
panic: reflect: nil type passed to Type.AssignableTo
goroutine 39 [running]:
reflect.(*rtype).AssignableTo(0x18d7ca0, 0x0, 0x0, 0xc0005be310)
	/usr/local/go/src/reflect/type.go:1445 +0xd3
github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInputAndDetermineSecret(0x18d7ca0, 0xc0005be310, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/justin/go/src/github.com/pulumi/pulumi/sdk/go/pulumi/rpc.go:206 +0x31a5
github.com/pulumi/pulumi/sdk/v3/go/pulumi.marshalInput(0x18d7ca0, 0xc0005be310, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/justin/go/src/github.com/pulumi/pulumi/sdk/go/pulumi/rpc.go:171 +0x73
github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).Call.func1(0x0, 0x0, 0x0)
	/Users/justin/go/src/github.com/pulumi/pulumi/sdk/go/pulumi/context.go:367 +0x236
github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).Call.func2(0xc0005d8000, 0x1954dda, 0x28, 0x1a36f70, 0xc0005be460, 0xc000112640)
	/Users/justin/go/src/github.com/pulumi/pulumi/sdk/go/pulumi/context.go:457 +0x122
created by github.com/pulumi/pulumi/sdk/v3/go/pulumi.(*Context).Call
	/Users/justin/go/src/github.com/pulumi/pulumi/sdk/go/pulumi/context.go:421 +0x225
```

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -17,6 +17,9 @@
 
 - [sdk/go] - Fix target and replace options for the Automation API.
   [#7426](https://github.com/pulumi/pulumi/pull/7426)
-  
+
 - [cli] - Don't escape special characters when printing JSON.
   [#7593](https://github.com/pulumi/pulumi/pull/7593)
+
+- [sdk/go] - Fix panic when marshaling `self` in a method.
+  [#7604](https://github.com/pulumi/pulumi/pull/7604)

--- a/sdk/go/pulumi/context.go
+++ b/sdk/go/pulumi/context.go
@@ -364,7 +364,7 @@ func (ctx *Context) Call(tok string, args Input, output Output, self Resource, o
 		// If we have a value for self, add it to the arguments.
 		if self != nil {
 			var deps []URN
-			resolvedSelf, selfDeps, err := marshalInput(self, anyType, true)
+			resolvedSelf, selfDeps, err := marshalInput(self, reflect.TypeOf(self), true)
 			if err != nil {
 				return nil, fmt.Errorf("marshaling __self__: %w", err)
 			}

--- a/sdk/go/pulumi/context.go
+++ b/sdk/go/pulumi/context.go
@@ -364,7 +364,7 @@ func (ctx *Context) Call(tok string, args Input, output Output, self Resource, o
 		// If we have a value for self, add it to the arguments.
 		if self != nil {
 			var deps []URN
-			resolvedSelf, selfDeps, err := marshalInput(self, nil, true)
+			resolvedSelf, selfDeps, err := marshalInput(self, anyType, true)
 			if err != nil {
 				return nil, fmt.Errorf("marshaling __self__: %w", err)
 			}

--- a/tests/integration/construct_component_methods/go/main.go
+++ b/tests/integration/construct_component_methods/go/main.go
@@ -72,6 +72,10 @@ func (o ComponentGetMessageResultOutput) Message() pulumi.StringOutput {
 	return o.ApplyT(func(v ComponentGetMessageResult) string { return v.Message }).(pulumi.StringOutput)
 }
 
+func (*Component) ElementType() reflect.Type {
+	return reflect.TypeOf((*Component)(nil))
+}
+
 func init() {
 	pulumi.RegisterOutputType(ComponentGetMessageResultOutput{})
 }

